### ISSUE TITLE
[Mobile] Cannot initialize InferenceSession on iOS net7.0-ios project

### ISSUE THREAD

**Issue by gench23** [microsoft/onnxruntime#18363]
### Describe the issue

When instantiating an InferenceSession I get this exception:
```
Attempting to JIT compile method '(wrapper managed-to-native) intptr Microsoft.ML.OnnxRuntime.NativeMethods/DGetMapKeyType:wrapper_aot_native (object,intptr,intptr&)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
```

### To reproduce

Just launch the application of the attached example to reproduce.
The project is built from the VS template.
[TestOnnxIos.zip](https://github.com/microsoft/onnxruntime/files/13309542/TestOnnxIos.zip)

### Urgency

Urgent, thanks for the help.

### Platform

iOS

### OS Version

15.4.1

### ONNX Runtime Installation

Released Package

### Package Name (if 'Released Package')

Microsoft.ML.OnnxRuntime

### ONNX Runtime Version or Commit ID

1.16.2

### ONNX Runtime API

C#

### Architecture

ARM64

### Execution Provider

Default CPU

**Comment by skottmckay** [comment]
Ugh. Sorry - I didn't realize there were multiple delegates that lacked the attribute. I checked all the changes in the PR that added DOrtGetResizedStringTensorElementBuffer (the previously reported error) and there were no other issues. There was an earlier PR that added more delegates that was overlooked, and DGetMapKeyType is in that. We need some more test infrastructure for these issues that require an actual device to be used.

I'll fix the other ones and add a script to check for mismatches between delegate and `UnmanagedFunctionPointer` attributes to try and catch it in the CI immediately so at least we have some coverage until we have better on-device test infrastructure. I can most likely publish a test nuget package with those changes if it helps, but most likely the complete fix won't be available until the next release (before EOY all going well).

**Comment by skottmckay** [comment]
Alternative would be to use the mono interpreter short term. Not sure if that's a viable option for you.

**Comment by gench23** [comment]
Thanks for the answer, unfortunately using the mono interpreter is not viable for me.
In order to have the fixes available, should I wait for the next release? Because I tried to build the nuget package using the commands in the Readme in the csharp folder but I met an error while building by doing so.
To build I did the following on a Windows 11 PC:
`msbuild -t:restore ./src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj -p:SelectedTargets=PreNet6`

`msbuild -t:build ./src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj -p:SelectedTargets=PreNet6`

`dotnet build ./src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj -p:SelectedTargets=Net6`

`msbuild -t:restore ./src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj -p:SelectedTargets=All`

`msbuild ./OnnxRuntime.CSharp.proj -t:CreatePackage -p:OrtPackageId=Microsoft.ML.OnnxRuntime -p:Configuration=Debug -p:Platform="Any CPU"`

And the error I get while building the previous project is this:
`/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/16.4.7125/targets/Xamarin.Shared.Sdk.targets(3,3): Error: clang++ exited with code 1:
Undefined symbols for architecture arm64:
  "_OrtGetApiBase", referenced from:
     -u command line option
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation) (SDEProfileReader.iOS)`

Did I miss something maybe?

**Comment by skottmckay** [comment]
The script in the repo tries to build the native and managed packages as it assumes a native build was done. It's more for local testing than real usage. 

I can run our nuget packaging pipeline with the change and publish the Microsoft.ML.OnnxRuntime.Managed package on the nuget testing repo (same as what I did with the dummy extensions package) if that helps. Should be able to reference that with the 1.16.2 Microsoft.ML.OnnxRuntime native package to pickup the fixes.

**Comment by gench23** [comment]
Ah ok, got it.

Yes, that would really help me in the process actually. Can you share the link to the nuget package when it will be available?

**Comment by gench23** [comment]
Also, if I wanted to build the nuget package on my own, how should I proceed? Looking at the documentation, the steps are not very clear. Knowing this would really help in my iterations, as I'd move along much faster I think.

**Comment by skottmckay** [comment]
Hopefully this one works: (https://int.nugettest.org/packages/Microsoft.ML.OnnxRuntime.Managed/1.17.0-dev-20231113-0108-c7f5d43bda)

You should be able to do a build of the native and managed packages with build.bat/build.sh using the `--build_nuget` arg.

**Comment by gench23** [comment]
Thanks for the reply, I downloaded the package and add it to the project, but now I get thi error at built time:
```
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/16.4.7125/targets/Xamarin.Shared.Sdk.targets(3,3): Error: clang++ exited with code 1:
Undefined symbols for architecture arm64:
  "_OrtGetApiBase", referenced from:
     -u command line option
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation) (TestOnnxIos)
```

I imported the package replacing the other one, but I noticed this one is only the Managed part, do I need to do something different?

**Comment by gench23** [comment]
Also, importing it alongside the one from nuget, like this:
```
<ItemGroup>
    <PackageReference Include="SkiaSharp" Version="2.88.6" />
    <PackageReference Include="Microsoft.ML.OnnxRuntime" Version="1.16.2" />
    <PackageReference Include="Microsoft.ML.OnnxRuntime.Managed" Version="1.17.0-dev-20231103-0203-016b75260b" />
</ItemGroup>
```
I happen to get the same error.

**Comment by skottmckay** [comment]
Not sure what's going on there. The symbol is defined in the native library. I extracted the `onnxruntime` library from here:

<img width="216" alt="image" src="https://github.com/microsoft/onnxruntime/assets/979079/97c10570-174a-4baf-ae39-826b7f4ca563">

and the symbol is defined and global

```
nm -gU onnxruntime | grep OrtGetApiBase
0000000000003c28 T _OrtGetApiBase
```

Does this potentially apply to your combination of tools? https://github.com/xamarin/xamarin-macios/issues/16223

**Comment by gench23** [comment]
My bad, I think I explained myself poorly. I meant that when I import the package you built for me alongside the one from nuget, the application builds, but I get the same runtime error I started with.

**Comment by gench23** [comment]
I confirm it was an error on my side, I don't know why, but when I imported the package from VS 2022 for Mac, it actually imported a different version I had tried to build myself while actually showing the one you had built in the nuget package manager, something very strange. Probably caching issues or something.

Anyway, the package you uploaded works and now I can create an InferenceSession and run it getting the expected result, thank you very much for the help @skottmckay 👍🏻

**Comment by skottmckay** [comment]
Excellent. Will be in the 1.17 release before EOY (all going well).

I've also published a more official version of the 'dummy' package to provide the RegisterCustomOps for now. https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime.Extensions.Dummy/0.10.0-beta. 'beta' as we haven't released 0.10 of the 'real' extensions package yet.
